File: 프로그래머스/Level3/디스크 컨트롤러.py
# https://programmers.co.kr/learn/courses/30/lessons/42627

import heapq

def solution(jobs):
    jobs.sort()
    min_heap = []
    answer = 0
    i, start, now = 0, -1, 0
    while i < len(jobs):
        for job in jobs:
            if start < job[0] <= now:
                heapq.heappush(min_heap, (job[1], job[0]))
        if len(min_heap) > 0:
            temp = heapq.heappop(min_heap)
            start = now
            now += temp[0]
            answer += (now - temp[1])
            i += 1
        else:
            now += 1
    return answer // len(jobs)

File: 프로그래머스/Level3/디스크 컨트롤러.js
// https://programmers.co.kr/learn/courses/30/lessons/42627

const solution = (jobs) => {
  let answer = 0,
    j = 0,
    time = 0;
  jobs.sort((a, b) => a[0] - b[0]);
  const q = [];

  while (j < jobs.length || q.length !== 0) {
    if (jobs.length > j && time >= jobs[j][0]) {
      q.push(jobs[j++]);
      q.sort((a, b) => a[1] - b[1]);
      continue;
    }
    if (q.length !== 0) {
      time += q[0][1];
      answer += time - q[0][0];
      q.shift();
    } else {
      time = jobs[j][0];
    }
  }
  return parseInt(answer / jobs.length);
};
